docs/client-sdk-states-and-lifecycle.mdx

```diff
@@ -0,0 +1,414 @@
+
+The [behaviour of API calls](#client-behaviors) to the Client SDK will
+depend on both the last observed [server-side experiment
+state](#server-side-experiment-states), and the current [local client
+state](#client-side-experiment-states), and the SDK will [emit
+telemetry](#telemetry) based on the local state of each
+experiment.
+
+I happen to think that these states could map nicely to some Rust Enums, but your mileage may vary.
+
+
+## Key Concepts
+
+
+#### Treatment
+
+We'll use "treatment" to refer to the change in client behaviour that's triggered by an experiment. This might be a small change in colour or wording, or may provide some much larger piece of new UX. The details and scope don't matter for our purposes here, what matters is having a word that means "the user is experiencing the effects of the experiment".
+
+
+#### Enrollment
+
+Enrollment is the process of the client SDK deciding whether this client should be part of the experiment and if so, what branch it should be assigned to. We'll emit an [enrollment](https://mana.mozilla.org/wiki/display/FJT/Nimbus+Engineering#NimbusEngineering-ExperimentTelemetry) telemetry event when this happens.
+
+Once a client is enrolled in an experiment, it will tag its telemetry pings with that experiment and branch information so that they can be identified for analysis.
+
+Enrolled clients will remain enrolled and assigned to the same branch until the experiment ends, or until they reset their experiment state by disabling telemetry. A client may _disable the treatment_ under various circumstances, but it's important that it continue reporting that it was enrolled in the experiment, to [help with data analysis](https://docs.google.com/document/d/18Vdn4zZX2D1u9AqzHRY0lUSUbpBdouqWtVp3Bu8vYQA/).
+
+Note that it's possible for an experiment to be closed for new enrollments, while continuing to be active for and showing its treatment to previously-enrolled clients.
+
+
+#### Exposure
+
+Exposure is the process of actually showing the treatment. In other words, it's when we change client behaviour based on the experiment branch in which the client was enrolled, or when we _would_ have changed behaviour if the user was not in the control group. We'll emit an [exposure](https://mana.mozilla.org/wiki/display/FJT/Nimbus+Engineering#NimbusEngineering-ExperimentTelemetry) telemetry event the first time this happens.
+
+Note that it's possible for a client to be enrolled in an experiment but never be exposed to its treatment, for example if the user never uses the part of the client app that is affected by the treatment.
+
+
+#### Disqualification
+
+After a client is enrolled in an experiment, we may need to disable that experiment's treatment for various reasons, such as:
+
+
+
+*   The user explicitly opts out of the experiment, or of all experiments.
+*   The application code detects that it cannot show the treatment.
```

Comment:
A comment thread from the original doc asks:

> (@rfk) TODO: how would the application signal this?
> (@mhammond) and can we get an example? At face value, that sounds like an issue with targeting, but I'm sure there's something subtle I'm missing.
>@rfk One example from previous discussions, is if the experiment treatment affects some default part of the UX, but the user goes into their settings and explicitly configures the behaviour of that part of the UX in a way that conflicts with the experiment (e.g. disabling "feature recommendations" and thus interfering with any experiment that tries to tweak feature recommendation text).

IIRC the current system solves for this by adding a bunch of targetting attributes for things like "has the user disabled feature recommendations?" but there are concerns about scalability and maintainability of this approach.

Reply:
Right now, the way this can happen is by users opting out via UI – there might be other cases we consider in the future, like a user-facing preference being changed

Reply:
I think we should strive to avoid sending an exposure event at all in this situation and rely on exposure instead of targeting to filter out clients that have the surface disabled. An "exposure failure" concept would require more complexity in Jetstream.

Reply:
https://github.com/mozilla/experimenter/issues/5164 filed to track this, marking resolved.